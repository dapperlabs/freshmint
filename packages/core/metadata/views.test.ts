import TemplateGenerator from '../generators/TemplateGenerator';
import * as metadata from './index';

const name = 'Foo NFT Collection';
const description = 'This is the Foo NFT collection.';
const url = 'http://foo.com';
const ipfsMedia = 'bafkreicrfbblmaduqg2kmeqbymdifawex7rxqq2743mitmeia4zdybmmre';
const httpMedia = 'http://foo.com/nft.jpeg';
const mediaType = 'image/jpeg';

const contractName = 'Foo';

function generateView(view: metadata.View): string {
  return TemplateGenerator.generate(view.type.cadenceTemplate, { contractName, view });
}

describe('NFTCollectionDisplayView', () => {
  it('rejects combined IPFS and HTTP media input', () => {
    expect(() => {
      metadata.NFTCollectionDisplayView({
        name,
        description,
        url,
        media: {
          ipfs: {
            cid: ipfsMedia,
          },
          url: httpMedia,
          type: mediaType,
        },
      });
    }).toThrow();
  });

  const ipfsView = metadata.NFTCollectionDisplayView({
    name,
    description,
    url,
    media: {
      ipfs: {
        cid: ipfsMedia,
      },
      type: mediaType,
    },
  });

  const ipfsWithPathView = metadata.NFTCollectionDisplayView({
    name,
    description,
    url,
    media: {
      ipfs: {
        cid: ipfsMedia,
        path: 'foo.jpeg',
      },
      type: mediaType,
    },
  });

  const compactIpfsView = metadata.NFTCollectionDisplayView({
    name,
    description,
    url,
    media: {
      ipfs: ipfsMedia,
      type: mediaType,
    },
  });

  const legacyIpfsView = metadata.NFTCollectionDisplayView({
    name,
    description,
    url,
    media: {
      ipfsCid: ipfsMedia,
      type: mediaType,
    },
  });

  it('generates a Cadence snippet with an IPFS media file with no path', () => {
    expect(generateView(ipfsView)).toMatchSnapshot();
  });

  it('generates a Cadence snippet with an IPFS media file with a path', () => {
    expect(generateView(ipfsWithPathView)).toMatchSnapshot();
  });

  it('generates a Cadence snippet with an IPFS media file passed as a CID string only', () => {
    expect(compactIpfsView).toEqual(ipfsView);
    expect(generateView(compactIpfsView)).toEqual(generateView(ipfsView));
  });

  it('generates a Cadence snippet with an IPFS media file using legacy input', () => {
    // Legacy view should be equivalent to non-legacy view
    expect(legacyIpfsView).toEqual(ipfsView);
    expect(generateView(legacyIpfsView)).toEqual(generateView(ipfsView));
  });

  it('generates a Cadence snippet with an HTTP media file', () => {
    const view = metadata.NFTCollectionDisplayView({
      name,
      description,
      url,
      media: {
        url: httpMedia,
        type: mediaType,
      },
    });

    expect(generateView(view)).toMatchSnapshot();
  });

  it('returns the collectionMetadata field when no options are passed', () => {
    const view = metadata.NFTCollectionDisplayView();
    expect(generateView(view)).toMatchSnapshot();
  });
});

describe('SerialView', () => {
  it('generates a Cadence snippet', () => {
    const serialNumberField = new metadata.Field('serialNumber', metadata.UInt64());
    const serialView = metadata.SerialView({ serialNumber: serialNumberField });

    expect(generateView(serialView)).toMatchSnapshot();
  });

  it('rejects a field that is not typed as UInt64', () => {
    const invalidSerialNumberField = new metadata.Field('serialNumber', metadata.Int());

    expect(() => metadata.SerialView({ serialNumber: invalidSerialNumberField })).toThrow(
      `The serialNumber field passed to SerialView must have type UInt64. You passed the 'serialNumber' field which has type Int.`,
    );
  });
});

describe('ExternalURLView', () => {
  it('accepts a complete Cadence expression', () => {
    const externalURLView = metadata.ExternalURLView({ cadenceTemplate: '"http://foo.com/nfts/".concat(self.id)' });
    expect(generateView(externalURLView)).toMatchSnapshot();
  });

  it('accepts a raw template that uses no variables', () => {
    const externalURLView = metadata.ExternalURLView('https://foo.com');
    expect(generateView(externalURLView)).toMatchSnapshot();
  });

  it('accepts a raw template that uses the ${collection.url} variable', () => {
    const externalURLView = metadata.ExternalURLView('${collection.url}/nfts');
    expect(generateView(externalURLView)).toMatchSnapshot();
  });

  it('accepts a raw template that uses the ${nft.owner} variable', () => {
    const externalURLView = metadata.ExternalURLView('http://foo.com/${nft.owner}');
    expect(generateView(externalURLView)).toMatchSnapshot();
  });

  it('accepts a raw template that uses the ${nft.id} variable', () => {
    const externalURLView = metadata.ExternalURLView('http://foo.com/${nft.id}');
    expect(generateView(externalURLView)).toMatchSnapshot();
  });

  it('accepts a raw template that uses all variables', () => {
    const externalURLView = metadata.ExternalURLView('${collection.url}/nfts/${nft.owner}/${nft.id}');
    expect(generateView(externalURLView)).toMatchSnapshot();
  });
});
