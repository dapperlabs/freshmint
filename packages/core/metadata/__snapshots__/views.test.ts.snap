// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalURLView accepts a complete Cadence expression 1`] = `
"pub fun resolveExternalURL(): MetadataViews.ExternalURL {
    return MetadataViews.ExternalURL(\\"http://foo.com/nfts/\\".concat(self.id))
}
"
`;

exports[`ExternalURLView accepts a raw template that uses all variables 1`] = `
"pub fun resolveExternalURL(): MetadataViews.ExternalURL {
    let collectionURL = Foo.collectionMetadata!.externalURL.url
    let nftOwner = self.owner!.address.toString()
    let nftID = self.id.toString()
    return MetadataViews.ExternalURL(collectionURL.concat(\\"/nfts/\\").concat(nftOwner).concat(\\"/\\").concat(nftID))
}
"
`;

exports[`ExternalURLView accepts a raw template that uses no variables 1`] = `
"pub fun resolveExternalURL(): MetadataViews.ExternalURL {
    return MetadataViews.ExternalURL(\\"https://foo.com\\")
}
"
`;

exports[`ExternalURLView accepts a raw template that uses the \${collection.url} variable 1`] = `
"pub fun resolveExternalURL(): MetadataViews.ExternalURL {
    let collectionURL = Foo.collectionMetadata!.externalURL.url
    return MetadataViews.ExternalURL(collectionURL.concat(\\"/nfts\\"))
}
"
`;

exports[`ExternalURLView accepts a raw template that uses the \${nft.id} variable 1`] = `
"pub fun resolveExternalURL(): MetadataViews.ExternalURL {
    let nftID = self.id.toString()
    return MetadataViews.ExternalURL(\\"http://foo.com/\\".concat(nftID))
}
"
`;

exports[`ExternalURLView accepts a raw template that uses the \${nft.owner} variable 1`] = `
"pub fun resolveExternalURL(): MetadataViews.ExternalURL {
    let nftOwner = self.owner!.address.toString()
    return MetadataViews.ExternalURL(\\"http://foo.com/\\".concat(nftOwner))
}
"
`;

exports[`NFTCollectionDisplayView generates a Cadence snippet with an HTTP media file 1`] = `
"pub fun resolveNFTCollectionDisplay(): MetadataViews.NFTCollectionDisplay? {
    let media = MetadataViews.Media(
        file: MetadataViews.HTTPFile(url: \\"http://foo.com/nft.jpeg\\"),
        mediaType: \\"image/jpeg\\"
    )

    return MetadataViews.NFTCollectionDisplay(
        name: \\"Foo NFT Collection\\",
        description: \\"This is the Foo NFT collection.\\",
        externalURL: MetadataViews.ExternalURL(\\"http://foo.com\\"),
        squareImage: media,
        bannerImage: media,
        socials: {}
    )
}
"
`;

exports[`NFTCollectionDisplayView generates a Cadence snippet with an IPFS media file with a path 1`] = `
"pub fun resolveNFTCollectionDisplay(): MetadataViews.NFTCollectionDisplay? {
    let media = MetadataViews.Media(
        file: MetadataViews.IPFSFile(
            cid: \\"bafkreicrfbblmaduqg2kmeqbymdifawex7rxqq2743mitmeia4zdybmmre\\", 
            path: \\"foo.jpeg\\"
        ),
        mediaType: \\"image/jpeg\\"
    )

    return MetadataViews.NFTCollectionDisplay(
        name: \\"Foo NFT Collection\\",
        description: \\"This is the Foo NFT collection.\\",
        externalURL: MetadataViews.ExternalURL(\\"http://foo.com\\"),
        squareImage: media,
        bannerImage: media,
        socials: {}
    )
}
"
`;

exports[`NFTCollectionDisplayView generates a Cadence snippet with an IPFS media file with no path 1`] = `
"pub fun resolveNFTCollectionDisplay(): MetadataViews.NFTCollectionDisplay? {
    let media = MetadataViews.Media(
        file: MetadataViews.IPFSFile(
            cid: \\"bafkreicrfbblmaduqg2kmeqbymdifawex7rxqq2743mitmeia4zdybmmre\\", 
            path: nil
        ),
        mediaType: \\"image/jpeg\\"
    )

    return MetadataViews.NFTCollectionDisplay(
        name: \\"Foo NFT Collection\\",
        description: \\"This is the Foo NFT collection.\\",
        externalURL: MetadataViews.ExternalURL(\\"http://foo.com\\"),
        squareImage: media,
        bannerImage: media,
        socials: {}
    )
}
"
`;

exports[`NFTCollectionDisplayView returns the collectionMetadata field when no options are passed 1`] = `
"pub fun resolveNFTCollectionDisplay(): MetadataViews.NFTCollectionDisplay? {
    return Foo.collectionMetadata
}
"
`;

exports[`SerialView generates a Cadence snippet 1`] = `
"pub fun resolveSerial(_ metadata: Metadata): MetadataViews.Serial {
    return MetadataViews.Serial(metadata.serialNumber)
}
"
`;
